bindings/python/tests/pire_test.py

```diff
@@ -0,0 +1,401 @@
+        )
+        check_scanner(
+            parse_scanner("(2.*)&([0-9]*_1+)", "a"),
+            accepts=["2123_1111", "2_1"],
+            rejects=["123_1111", "2123_1111$", "^_1"],
+        )
+        check_scanner(
+            parse_scanner("a|b|c"),
+            accepts=["a", "b", "c"],
+            rejects=["", "ab", "ac", "bc", "aa", "bb", "cc"],
+        )
+
+    def test_lexer_raises_on_parsing_invalid_regexp(self):
+        pytest.raises(Exception, pire.Lexer("[ab").Parse)
+
+    def test_lexer_accepts_unicode_pattern(self, parse_scanner):
+        check_scanner(
+            parse_scanner(u"юникод", "u"),
+            accepts=["\xd1\x8e\xd0\xbd\xd0\xb8\xd0\xba\xd0\xbe\xd0\xb4"],
+            rejects=["\xd1", ""],
+        )
+
+
+class TestScanner(object):
+    def test_scanner_inherits_from_base_scanner(self, scanner_class):
+        assert issubclass(scanner_class, pire.BaseScanner)
+
+    def test_state_inherits_from_base_state(self, scanner_class):
+        assert issubclass(scanner_class.StateType, pire.BaseState)
+
+    def test_state_type_property_is_set_right(self, scanner_class):
+        assert isinstance(scanner_class().InitState(), scanner_class.StateType)
+
+    def test_scanner_is_default_constructible(self, scanner_class):
+        scanner = scanner_class()
+        assert scanner.Empty()
+        if scanner_class is not pire.SlowScanner:
+            assert 1 == scanner.Size()
+        check_scanner(scanner, rejects=["", "some"])
+
+    def test_scanner_raises_when_matching_not_string_but_stays_valid(self, example_scanner):
+        for invalid_input in [None, False, True, 0, 42]:
+            pytest.raises(Exception, example_scanner.Matches, invalid_input)
+        check_scanner_is_like_example_scanner(example_scanner)
+
+    def test_scanner_is_picklable(self, example_scanner):
+        packed = pickle.dumps(example_scanner)
+        unpacked = pickle.loads(packed)
+        check_scanner_is_like_example_scanner(unpacked)
+
+    def test_scanner_is_saveable_and_loadable(self, example_scanner):
+        packed = example_scanner.Save()
+        unpacked = example_scanner.__class__.Load(packed)
+        check_scanner_is_like_example_scanner(unpacked)
+
+    def test_scanner_raises_when_loading_from_invalid_data(self, scanner_class):
+        invalid_data = "invalid"
+        pytest.raises(Exception, scanner_class.Load, invalid_data)
+
+        saved = scanner_class().Save()
+        pytest.raises(Exception, scanner_class.Load, saved[1:])
+
+    def test_scanner_finds_prefixes_and_suffixes(self, scanner_class):
+        fsm = pire.Lexer("-->").Parse()
+        any_occurence = scanner_class(~pire.Fsm.MakeFalse() + fsm)
+        first_occurence = scanner_class(~fsm.Surrounded() + fsm)
+        reverse_occurence = scanner_class(fsm.Reverse())
+
+        text = "1234567890 --> middle --> end"
+        assert 14 == first_occurence.LongestPrefix(text)
+        assert 11 == reverse_occurence.LongestSuffix(text[:14])
```

Comment:
Could you provide an example for which pytest differentiates between expected and observed? In simple cases it doesn't.
```
>       assert Say5() == 6
E       assert 5 == 6
E        +  where 5 = Say5()
```
```
>       assert 6 == Say5()
E       assert 6 == 5
E        +  where 5 = Say5()
```
```
>       assert Say5() == Say6()
E       assert 5 == 6
E        +  where 5 = Say5()
E        +  and   6 = Say6()
```